1-feature-1-task-list

chazzybearr opened the pull request.

chazzybearr committed:
Add unit tests for Tag tests
Fix method name in TagInteractor.java

nitvishn reviewed (dismissed):
Looks fine.

chazzybearr committed:
Merge branch 'main' of https://github.com/CSC207-2022F-UofT/course-project-productive-potato-sloth into 1-feature-1-task-list

chazzybearr committed:
Add factory classes for Tag and Task
Update Tag and Task to implement Serializable

chazzybearr committed:
Add TagDataAccessInterface.java, TagDatabaseGateway.java, TagInputBoundary.java, TagRequestModel.java, TagResponseModel.java

chazzybearr committed:
Merge branch 'main' of https://github.com/CSC207-2022F-UofT/course-project-productive-potato-sloth into 1-feature-1-task-list

Dawei-He2002 commented on src/test/java/useCases/Tags/TestCreateTag.java:
```diff
@@ -0,0 +1,53 @@
+     *
+     * @param name The name of the Tag
+     */
+    @ParameterizedTest
```
Good way to test classes and methods with parameters. I might use this way of running tests for my usecase and controller classes.

Dawei-He2002 commented on src/test/java/useCases/Tags/TestEditTag.java:
```diff
@@ -0,0 +1,101 @@
+         */
+        @ParameterizedTest
+        @ValueSource(strings = {"tag", "t", ""})
+        public void testEditColorDifferent(String name) {
```
Very comprehensive testing of all use cases.

Dawei-He2002 commented on src/main/java/entities/TaskFactory.java:
```diff
@@ -0,0 +1,13 @@
+        return new Task(name, user);
+    }
+    
+    public Task create(String name, User user, String description) {
```
I have a question: is the factory really necessary if all it does is call the corresponding method on the Task entity? There is only one type of Task, so would it not be more convenient to directly call the initialization method in Task?

nitvishn replied:
I think it's good design.

Dawei-He2002 commented on src/test/java/useCases/Tags/TestDeleteTag.java:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    /**
+     * Delete Tag not yet implemented
```
This is not yet implemented because it involves the database, right?

Dawei-He2002 reviewed (dismissed):
Overall, nothing wrong here. I'd say it's ready to merge if Delete Tag manages to be implemented.

Dawei-He2002 commented on src/main/java/entities/Tag.java:
```diff
@@ -18,24 +24,26 @@ public class Tag {
+     *              // @param user The owner of the Tag
      */
-    public Tag(String name, Color color /* User user */) {
+    public Tag(String name, Color color, User user) {
```
I haven't reviewed the code from the "autograding" perspective last time, but on the autograde test this threw an Error because the CreateTag class called this method without the User parameter. One solution would be to implement an overloaded method with the User set to null and assign a User later, or otherwise to pass a User into CreateTag (either as an instance variable or a param) and do it without overloading.

Dawei-He2002 commented on src/main/java/gateways/Tags/TagDatabaseGateway.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Override
+    public int update(Tag tag) {
```
For update and delete, both threw errors because the DataAccessInterface required them to return booleans instead of integers. Why do they return integers here anyways? If the output must be integers, having them return false/true and then having a line of code in the method that handles them convert that to the desired integer (or maybe having a helper method) would probably take care of the error.

Dawei-He2002 commented on src/main/java/gateways/Tags/TagDatabaseGateway.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Override
+    public int insert(Tag tag) {
```
The same problem exists for insert, compared to delete. On the interface it returns void. Maybe consider modifying the data access interface if no other people use it.

Dawei-He2002 commented on src/main/java/gateways/Tags/TagDatabaseGateway.java:
```diff
@@ -0,0 +1,77 @@
+    }
+
+    @Override
+    public Tag get(int id) {
```
Autograding threw errors for all @Override methods except for getAll, because it did not "implement or override a corresponding supertype method". I think for inser, delete and edit these might be caused by the autograder bugging out due to the existing errors, but it is true that TagDataAccessInterface does not have a get method, so the Override is unnecessary (as of this version).

Dawei-He2002 reviewed (commented):
Added feedback to autograding errors as well as some general advice on fixing them.

chazzybearr committed:
Add input boundaries for each use case
Add Request and Response models for Tag and Task
Update use cases to parse Request model data and to return Response model
Consolidate MarkTaskComplete.java and MarkTaskIncomplete.java into MarkTaskCompleteness.java
Add Database Gateways for Tag and Task
Add tests for Task and Tag Factory
Remove TaskInteractor.java and its corresponding tests
Remove test classes for use cases (will be added later)
Uncomment certain lines from User

chazzybearr committed:
Merge branch 'main' of https://github.com/CSC207-2022F-UofT/course-project-productive-potato-sloth into 1-feature-1-task-list

chazzybearr committed:
Change unmappable characters in TimerInputData.java and TimerInteractor.java

chazzybearr committed:
Change unmappable characters in TimerInputData.java and TimerInteractor.java

chazzybearr committed:
Remove TestTagDatabaseGateway.java
Relies on User.tags being instantiated to an empty list, but is currently null.

chazzybearr committed:
Update use cases to take the Data Access Interface rather than the implementing class to follow clean architecture
Fix typo in use cases

nitvishn commented on src/main/java/entities/TagFactory.java:
```diff
@@ -0,0 +1,11 @@
+
+import java.awt.Color;
+
+public class TagFactory {
```
Good use of the Factory class.

nitvishn commented on src/main/java/entities/Tag.java:
```diff
@@ -69,7 +80,16 @@ public void setColor(Color color) {
-//    public User getUser() {
-//        return this.user;
-//    }
+    public User getUser() {
```
Ideally this change should have been made by Chhavi, but seeing that it was made in correspondence with her, it seems appropriate.

nitvishn commented on src/main/java/entities/Task.java:
```diff
@@ -1,12 +1,12 @@
  * An entity class representing a Task
  */
-public class Task {
+public class Task implements Serializable {
```
This should change -- we only need User to be serializable as we are only storing User.

nitvishn commented on src/main/java/entities/Task.java:
```diff
@@ -35,17 +35,17 @@ public class Task {
      * A list of the associated events of this Task
      */
-//    ArrayList<Event> events;
+    ArrayList<Event> events;
```
Please use List instead of ArrayList.

nitvishn commented on src/main/java/entities/Task.java:
```diff
@@ -165,56 +165,61 @@ public Boolean getCompleted() {
+     *
+     * @return events A list of events associated with this Task
+     */
+    public ArrayList<Event> getEvents() {
```
Please use List instead of ArrayList.

nitvishn commented on src/main/java/entities/Task.java:
```diff
@@ -165,56 +165,61 @@ public Boolean getCompleted() {
+     *
+     * @return List of the collaborators of this task
+     */
+    public ArrayList<User> getCollaborator(User collaborator) {
```
Here too.

nitvishn commented on src/main/java/gateways/Tags/TagDatabaseGateway.java:
```diff
@@ -0,0 +1,103 @@
+    }
+
+    public void load() {
+        this.userList = userDatabaseGateway.loadFromFile();
```
You should use getAll, loadFromFile shouldn't be called outside the class. I should have made it private, sorry about that.

nitvishn commented on src/main/java/gateways/Tags/TagDatabaseGateway.java:
```diff
@@ -0,0 +1,103 @@
+    private UserDatabaseGateway userDatabaseGateway;
+    private List<User> userList;
+    private User currentUser;
+    private ArrayList<Tag> tagList;
```
Please use List.

nitvishn commented on src/main/java/gateways/Tags/TagDatabaseGateway.java:
```diff
@@ -0,0 +1,103 @@
+public class TagDatabaseGateway extends DatabaseGateway implements TagDataAccessInterface {
+
+    private final CurrentUserService currentUserService;
+    private UserDatabaseGateway userDatabaseGateway;
```
The type here should be DataAccessInterface<User>.

nitvishn replied:
I meant `DataAccessInterface<User>`

nitvishn commented on src/main/java/gateways/Tags/TagDatabaseGateway.java:
```diff
@@ -0,0 +1,103 @@
+
+    public void load() {
+        this.userList = userDatabaseGateway.loadFromFile();
+        this.currentUser = userDatabaseGateway.get(currentUserService.getCurrentUser().getUsername());
```
I believe the currentUserService has a `getUser` method.

nitvishn commented on src/main/java/gateways/Tasks/TaskDatbaseGateway.java:
```diff
@@ -0,0 +1,103 @@
+    private UserDatabaseGateway userDatabaseGateway;
+    private List<User> userList;
+    private User currentUser;
+    private ArrayList<Task> taskList;
```
Code smelly! Make it a List

nitvishn commented on src/main/java/gateways/Tasks/TaskDatbaseGateway.java:
```diff
@@ -0,0 +1,103 @@
+
+
+    private final CurrentUserService currentUserService;
+    private UserDatabaseGateway userDatabaseGateway;
```
Again, need to have an instance of DataAccessInterface<User>.

chazzybearr replied:
The DataAccessInterface currently doesn't have a save method, so I have no way of persisting the data once changes are made. I'll remember to make these changes once this has been added (I think you also mentioned this during our meeting today).

nitvishn commented on src/main/java/gateways/Tasks/TaskDatbaseGateway.java:
```diff
@@ -0,0 +1,103 @@
+            userDatabaseGateway.saveToFile();
+            return false;
+        }
+        userDatabaseGateway.saveToFile();
```
We'll change this soon by having a saveData method in the interface DataAccessInterface that serves this purpose.

nitvishn commented on src/main/java/useCases/Tags/DeleteTag.java:
```diff
@@ -1,14 +1,39 @@
+public class DeleteTag implements DeleteTagInputBoundary {
 
-        // Requires data access object to be inplemented
+    private final TagDataAccessInterface databaseGateway;
```
Did you mean `DataAccessInterface<Tag>`?

nitvishn replied:
Nevermind.

nitvishn commented on src/main/java/useCases/Tags/EditTag.java:
```diff
@@ -1,41 +1,56 @@
+public class EditTag implements EditTagInputBoundary {
 
-    Tag tag;
+    TagDataAccessInterface databaseGateway;
```
Again, did you mean DataAccessInterface<Tag>?

nitvishn commented on src/main/java/gateways/Tasks/TaskResponseModel.java:
```diff
@@ -0,0 +1,20 @@
+package gateways.Tasks;
+
+public class TaskResponseModel {
```
Having the class name TaskResponseModel is slightly confusing, maybe have it reflect the use case. For instance, CreateTaskResponseModel might be a little more appropriate.

chazzybearr replied:
I've decided to use one request and response model for Task and Tag. Since I have a significant number of use cases, I feel it would be a lot of redundant code to create request and response models for each. Thus, I have decided to consolidate all the data required for all the use cases into one class.

nitvishn commented on src/test/java/entities/TaskFactoryTest.java:
```diff
@@ -0,0 +1,49 @@
+     */
+    @ParameterizedTest
+    @ValueSource(strings = {"tag", "t", ""})
+    public void testTaskName(String name) {
```
Good tests.

nitvishn reviewed (changes requested):
20 comments, Leon :) 20 comments :)

chazzybearr committed:
Addressed all requested changes
Fix class name typo in TaskDatabaseGateway.java

chazzybearr committed:
Merge main into 1-feature branch

chazzybearr committed:
Empty commit

chazzybearr commented:
The empty commit was strange - git was complaining that "You have not concluded your merge" and someone on StackOverflow said to make an empty commit by literally committing nothing using the terminal and that seemed to fix the issue (??)

chazzybearr committed:
Isolated DeleteTaskConfirmationDialog and DeleteTaskPopUp into their own classes
Fix bug where UI would freeze after making adding/deleting/editing a task

chazzybearr commented:
I've extracted two classes - the popup and confirmation dialogue - to remove bloaters in my screens.

chazzybearr commented:
My features seem to be mostly finished, I'll start refactoring and polishing soon.

chazzybearr committed:
Connected Task List screen to the View Tasks button on the welcome page

chazzybearr commented:
I've changed the main file to instantiate all my controllers, interactors, and screens. 
The task list screen is now connected to a button on the welcome page!
As my user story is quite expansive (12 controllers, 12 use cases, 6 screens), I've had to instantiate every one of them in main, causing a bit of clutter. 
I've also tried my best to comment on what each code section does as clearly as possible to avoid any confusion.

chazzybearr commented:
For some design choices I made:

> I chose to have only one TaskRequestModel and TaskResponse model class, rather than having request and response model classes for each use case. 
> 
> As stated above, with 12 use cases, I would have to have an additional 24 classes (one TaskRequest and one TaskResponse). 
> 
> For a program of this size, I concluded it would be unnecessary to have such a titanic quantity of lazy classes. 
> 
> Thus, I made the choice to combine all the request and response models into one class each. 
>
> Though this does break clean architecture in the Single Responsibility Principle, I concluded that if this program were to expand, it would be trivial to break the generic TaskRequest and Response models for their separate use cases. 
>
> I would be very open to refactoring this with some discussion

> Same comments and reasoning with the TagResponse and Request Models

chazzybearr committed:
Fix bug where adding task did not persist data
Add serializable for Tag and Task

chazzybearr commented:
Tag was throwing a NotSeriazableException on runtime, so I made Tag and Task implement Serializable.

chazzybearr committed:
Add Javadoc for every class

chazzybearr commented:
After painstakingly going through all my files, I have added Javadoc and commented my classes as much as possible.

chazzybearr commented:
I chose to only add Javadoc to public methods, and skip Javadoc for private methods, as these methods are not being called by any external class, so I felt it was unnecessary. (There are only like 4 private methods anyways.)

Dawei-He2002 commented on src/main/java/entities/Task.java:
```diff
@@ -224,6 +224,11 @@ public List<User> getCollaborator() {
+    /**
+     * Gets the Chatroom associated with this Task
+     *
+     * @return The Chatroom for this Task
```
Good thing that this is now implemented.

Dawei-He2002 reviewed (dismissed):
Extensive documentation. Code is very readable.

chazzybearr committed:
Add tests for most use cases and controllers

chazzybearr committed:
Pulled changes from origin main
